lib/convertor.js

```diff
@@ -98,61 +98,96 @@ class MongoToKnex {
-
-        /**
-         * We group the relations by a unique key.
-         * Each grouping will create a sub query.
-         *
-         * @TODO: choose a different key name (!)
-         */
-        _.each(relations, (relation) => {
-            if (!groupedRelations[relation.table]) {
-                groupedRelations[relation.table] = [];
-            }
-
-            if (groupedRelations[relation.table].length) {
-                const columnExists = _.find(groupedRelations[relation.table], (statement) => {
-                    /**
-                     * CASE:
-                     * - we should not use the same sub query if the column name is the same (two sub queries)
-                     * - e.g. $and conjunction requires us to use 2 sub queries, because we have to look at each individual tag
-                     */
-                    if (statement.column === relation.column) {
+    groupRelationStatements(statements, mode) {
+        const group = {};
+
+        // groups depend on the mode of grouping, if its and $and we need to treat a filter on
+        // joining table differently than we would with $or
+        // e.g. for $or we can create a subquery or group that filter,
+        //      for $and we have to include joining table filter in every group
+        const innerWhereStatements = (mode === '$and')
+            ? statements.filter(r => !(r.join_table))
+            : statements;
+
+        _.each(innerWhereStatements, (statement) => {
+            /**
+             * CASE:
+             * - we should not use the same sub query if the column name is the same (two sub queries)
+             * - e.g. $and conjunction requires us to use 2 sub queries, because we have to look at each individual tag
+             *
+             * - we should also not use grouping of negated values for the same reasons as above
+             */
+            let createSubGroup = isNegationOp(statement.operator);
+
+            if (!createSubGroup && group[statement.table]) {
+                createSubGroup = _.find(group[statement.table].innerWhereStatements, (innerStatement) => {
+                    if (innerStatement.column === statement.column) {
                         return true;
                     }
                 });
+            }
 
-                if (columnExists) {
-                    const newKey = `${relation.table}_${Math.floor(Math.random() * 100)}`;
-                    if (!groupedRelations[newKey]) {
-                        groupedRelations[newKey] = [];
-                    }
+            let groupKey = statement.table;
 
-                    groupedRelations[newKey].push(relation);
-                    return;
-                }
+            if (createSubGroup) {
+                // TODO: while this is 'good enough' approach this needs to be approached with something
+                // more stable than random from 1..100
+                groupKey = `${statement.table}_${Math.floor(Math.random() * 100)}`;
```

Comment:
This is a bit ugly left over from the previous approach to grouping. Didn't touch it for now to avoid bloating this PR. Now that grouping method is extracted, seems like this method scoped counter for grouping can be introduced to tackle this problem better

Reply:
Suggest raising an issue to fix this later but soon - seems like a bug waiting to happen 🐞

Reply:
nice hm?

https://github.com/NexesJS/mongo-knex/pull/18/files#diff-8ed5262100dab197d3b7fa273ca3e105R111

^ this is the TODO. We could link the [new issue](https://github.com/NexesJS/mongo-knex/issues/21) in the code base 👍